vtex/faststore
# Contextual pipeline for adding, updating and removing items

## What's the purpose of this pull request?

This pull request allows the Checkout API to skip some steps in its pipeline during an item quantity change on item operations, which increases the site performance and reduces unnecessary back-end requests and processes. A similar change was made for the IO stores in the following PRs: https://github.com/vtex-apps/minicart/pull/269, https://github.com/vtex-apps/order-items/pull/41, https://github.com/vtex-apps/checkout-resources/pull/71, https://github.com/vtex-apps/checkout-graphql/pull/125.

More specifically, the skipped step is the one that updates the orderForm `paymentData`, and although this is crucial information, we don't need to worry about this during the shopping. Currently, we update this data when entering the cart and checkout pages, as seen in the PR for the legacy checkout (https://github.com/vtex/vtex.js/pull/72) and for the IO checkout as well (https://github.com/vtex-apps/order-manager/pull/52).

These optimizations will affect the following components:
- the "add to cart" button
- the minicart "change item quantity" buttons
- the minicart "delete item" button

## How it works?

A new parameter `allowedOutdatedData: [String!]` needs to be added to the mutation `UpdateItemsMutation` and `AddToCartMutation`. In order to use this new parameter, it sends the `[ "paymentData" ]` value on its respective controllers (`updateItems` and `addToCart`).

## How to test it?

https://github.com/vtex-sites/carrefourbrfood.store/pull/394
https://sfj-23a04b5--carrefourbrfood.preview.vtex.app/

1. Go to [this product page](https://sfj-23a04b5--carrefourbrfood.preview.vtex.app/whisky-johnnie-walker-red-label-500-ml-9443827/p?skuId=6970)
2. Click on the "add to cart" button (the minicart should open)
4. Go [this GraphiQL](https://jamstackprod--carrefourbrfood.myvtex.com/admin/graphql-ide)
5. Select `vtex.checkout-graphql@0.60.0`
6. Paste and run this query `{ orderForm(orderFormId: "your orderform here") { value paymentData { installmentOptions { value } } } }`
5. Assert that `installmentOptions` is an empty array
6. Go to cart
7. Re-run the query
8. Assert that `installmentOptions` isn't empty

## Store PRs

https://github.com/vtex-sites/carrefourbrfood.store/pull/394
https://github.com/vtex-sites/marinbrasil.store/pull/472
https://github.com/vtex-sites/btglobal.store/pull/457
https://github.com/vtex-sites/storecomponents.store/pull/786
